### PR TITLE
Base type-to-file allocation on the ReferenceGraph code

### PR DESCRIPTION
**Pull request by Porges**
This simplifies it greatly and it no longer gets “stuck” on my WIP branch.

### CODE REVIEW

File: hack/generator/pkg/astmodel/package_definition.go [hunk 1]
```diff
@@ -84,131 +77,37 @@ func emitFiles(filesToGenerate map[string][]TypeDefinition, outputDir string) er
...
 
-		allocatedFileName := allocateFileName(filesReferencingType, typeToAllocate, hasPendingReferences)
-		if allocatedFileName == "" && skippedTypes > skipLimit {
-			// We've stalled during allocation of types, implying we have a cycle of mutually referencing types
-			allocatedFileName = allocateFilenameWhenStalled(typesToAllocate, filesToGenerate, len(filesReferencingType), typeToAllocate)
+	// rootsFor maps a type name to the roots that can reach it
+	rootsFor := make(map[TypeName][]TypeName)
+	for root := range graph.roots {
+		collected := make(TypeNameSet)
+		graph.CollectTypes(root, collected)
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/package_definition.go:87]: That method is public here? So maybe it does need to be public?

Comment (Porges) [reply hack/generator/pkg/astmodel/package_definition.go:87]: it’s in the same package

File: hack/generator/pkg/astmodel/reference_graph.go [hunk 2]
```diff
@@ -44,27 +44,39 @@ func NewReferenceGraphWithResourcesAsRoots(types Types) ReferenceGraph {
...
 // collectTypes returns the set of types that are reachable from the given
 // 'node' (and places them in 'collected').
-func (c ReferenceGraph) collectTypes(node TypeName, collected TypeNameSet) {
-	if collected.Contains(node) {
+func (c ReferenceGraph) collectTypes(depth int, node TypeName, collected ReachableTypes) {
+	if currentDepth, ok := collected[node]; ok {
 		// We can stop here - we've already visited this node.
+		// But first, see if we are at a lower depth:
+		if depth < currentDepth {
+			collected[node] = depth
```
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/reference_graph.go:72]: If you lower the depth of `node` then you need to keep processing, and the depths of the things reachable from `node` need to be lowered too.

File: hack/generator/pkg/astmodel/type_definition.go [hunk 3]
```diff
@@ -90,8 +90,8 @@ func (std *TypeDefinition) RequiredImports() []PackageReference {
 	return std.theType.RequiredImports()
 }
 
-// FileNameHint returns what a file that contains this definition (if any) should be called
+// FileNameHint returns what a file that contains this name (if any) should be called
 // this is not always used as we might combine multiple definitions into one file
```
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/type_definition.go:94]:
```suggestion
// this is not always used as we often combine multiple definitions into one file
```

File: hack/generator/pkg/astmodel/reference_graph.go (hunk 2 above)
Comment (Porges) [reply hack/generator/pkg/astmodel/reference_graph.go:72]: 🤔

Comment (Porges) [reply hack/generator/pkg/astmodel/reference_graph.go:72]: added a test case and fix

**Pull request merged**